src/test/java/HeapTest.java

```diff
@@ -0,0 +1,433 @@
+                .addEqualityGroup(unequalDifferentComparator)
+                .testEquals();
+    }
+
+
+    @Test
+    void add_nonComparableType_throwsClassCastException() {
+        Heap<Color> trouble = new Heap<>();
+        trouble.add(new Color(0));
+        assertThatExceptionOfType(ClassCastException.class).isThrownBy(() -> trouble.add(new Color(0)));
+    }
+
+    @Nested
+    class DefaultConstructor {
+
+        @BeforeEach
+        void createClassUnderTest() {
+            classUnderTest = new Heap<>();
+            preState = new Heap<>();
+        }
+
+        @Nested
+        class WhenEmpty {
+
+            @Test
+            void add_empty_returnsTrue() {
+                assertThat(classUnderTest.add(99)).isTrue();
+            }
+
+            @Test
+            void remove_empty_throwsNoSuchElementException() {
+                assertThatExceptionOfType(NoSuchElementException.class).isThrownBy(classUnderTest::remove);
+            }
+
+            @Test
+            void peek_empty_throwsNoSuchElementException() {
+                assertThatExceptionOfType(NoSuchElementException.class).isThrownBy(classUnderTest::peek);
+            }
+
+            @Test
+            void size_empty_returnsZero() {
+                assertThat(classUnderTest.size()).isEqualTo(0);
+            }
+
+            @Test
+            void isEmpty_empty_returnsTrue() {
+                assertThat(classUnderTest.isEmpty()).isTrue();
+            }
+
+            @Test
+            void toString_empty_returnsString() {
+                assertThat(classUnderTest.toString()).isEqualTo("[]");
+            }
+
+            @Nested
+            class WhenSingleton {
+
+                @BeforeEach
+                void addSingleton() {
+                    classUnderTest.add(10);
+                    preState.add(10);
+                }
+
+                @Test
+                void add_singleton_returnsTrue() {
+                    assertThat(classUnderTest.add(99)).isTrue();
+                }
+
+                @Test
+                void remove_singleton_returnsElement() {
+                    assertThat(classUnderTest.remove()).isEqualTo(10);
+                }
+
+                @Test
+                void remove_singleton_emptyHeap() {
+                    classUnderTest.remove();
+                    assertThat(classUnderTest).isEqualTo(new Heap<>());
+                }
+
+                @Test
+                void peek_singleton_returnsElement() {
+                    assertThat(classUnderTest.peek()).isEqualTo(10);
+                }
+
+                @Test
+                void peek_singleton_unchanged() {
+                    classUnderTest.peek();
+                    assertThat(classUnderTest).isEqualTo(preState);
+                }
+
+                @Test
+                void size_singleton_returnsSize() {
+                    assertThat(classUnderTest.size()).isEqualTo(1);
+                }
```

Comment:
May make sense to also assert that `size`, `isEmpty` etc are not mutating operations